Allow for fatal_oom detection by regex also when exit_code tests are present

bernt-matthias opened the pull request:
The current behavior of the stdio error handling is to apply regexes only if no (fatal) exit code was found: 

https://github.com/galaxyproject/galaxy/blob/c78a23cb873cf5cbcf177cfa7052ae6dad506eea/lib/galaxy/jobs/output_checker.py#L68

Thus, because fatal_oom > fatal currently fatal_oom can be detected if and only if
- they are indicated by an exit_code
- indicated by output to stderr/stdio but no other (fatal) exit_code rule is found.

This assumes that programmers use exit_codes to indicate different error conditions -- which is in my opinion a very optimistic assumption that virtually never applies (I can not remember a single program). C programmers, for instance, usually just use libc's EXIT_SUCCESS and EXIT_FAILURE macros and output a more specific message to stderr. Hence in my opinion exit codes are in practice (unfortunately) unspecific. 

Furthermore, the current implementations of galaxy tools test for non zero exit code only anyway (I grepped in all iuc,  devteam, and bgruening tools and only found a single one). 

The PR is changes this behavior by checking for regexes also if a FATAL exit_code was found (and also all specified regexes are tested). 

Furthermore: 
- As a suggestion I implemented the test line wise in reverse order. Intuitively I would expect this to be faster. The original code is still available as additional function - so this is easy to revert. 
- I introduced a function for checking a regex against a stream (which was a TODO in the code).

bernt-matthias committed:
- check regexp also if FATAL was detected by exit_code
- introduced functions for checking regex against stream
- check regex line by line instead of on the complete stream
  (the original behavior is coded in check_output_regex which
  could be removed if linewise is OK)

bernt-matthias committed:
also check all regexes

bgruening commented:
@bernt-matthias can you please also add a test tool or enhance one of the existing tools https://github.com/galaxyproject/galaxy/tree/dev/test/functional/tools

bgruening commented:
```
galaxy.jobs.output_checker: WARNING: Tool check encountered unexpected exception; assuming tool was successful: Traceback (most recent call last):
  File "/home/travis/build/galaxyproject/galaxy/lib/galaxy/jobs/output_checker.py", line 123, in check_output
    rexmsg, max_error_level = check_output_regex_byline(regex, stderr, max_error_level)
  File "/home/travis/build/galaxyproject/galaxy/lib/galaxy/jobs/output_checker.py", line 42, in check_output_regex_byline
    log.info("Job %s: %s" % (job.get_id_tag(), rexmsg))
NameError: global name 'job' is not defined
```

bernt-matthias committed:
added test and changed modification of stdout/err

before stdout and stderr were modified while testing for exit codes and
regexes. this had the potential problem that later tests also checked
this output which might lead to confusions.

furthermore a marker line has been introduces which differentiates
Galaxy and original tool output

bernt-matthias commented:
Done :) Bug has been fixed. Furthermore my latest commit changes the modification of stdout and stderr: 

Before stdout and stderr were modified while testing for exit codes and regexes. this had the potential problem that later tests also checked this output which might lead to confusions.

Furthermore a marker line has been introduces which differentiates Galaxy and original tool output.

nsoranzo commented on test/functional/tools/detect_errors.xml:
```diff
@@ -0,0 +1,140 @@
+            <assert_stdout>
+                <has_line line="Warning: Branch A was taken in execution"/>
+            </assert_stdout>
+(.venv) ➜  tools git:(topic/regex-for-oom) ✗ 
```
Copy/paste mistake?

bernt-matthias replied:
Should be fine: 
- program writes "Branch A seems to be of iterest .. maybe its related to the error?" to stdout
- tool adds "Warning: Branch A was taken in execution" to stdout, and thats what I'm testing for here. 

But maybe I'm missing something?

nsoranzo replied:
Look at https://travis-ci.org/galaxyproject/galaxy/jobs/393608375#L1180 ,  you cannot put text inside `<test>`, only tags.

bernt-matthias replied:
jep

nsoranzo commented:
@bernt-matthias Can you also take a look at the failed jobs at https://travis-ci.org/galaxyproject/galaxy/builds/393608367 ?

bernt-matthias committed:
fixed linting and copy paste errors

nsoranzo commented:
There a framework test failing: https://jenkins.galaxyproject.org/job/docker-framework/11857/testReport/functional.test_toolbox/TestForTool_unicode_stream/test_tool_000001/

Restarting didn't help.

bernt-matthias commented:
According to https://jenkins.galaxyproject.org/job/docker-framework/11857/console this seems to be related to the line `### END of messages added by Galaxy AND START of original stderr` that I added to separate the original streams and the messages added by the tool. 

What can I do about this?

jmchilton commented:
Can we just drop that addition? How about drop that addition and we will backport this and I'll open a PR against dev that adds a new field on ``Job`` like ``failure_reason`` so we don't have to modify the input and output detected?

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into topic/regex-for-oom

jmchilton commented:
I merged this with dev and pushed into your branch - I hope that is okay.

jmchilton commented:
> As a suggestion I implemented the test line wise in reverse order. Intuitively I would expect this to be faster. The original code is still available as additional function - so this is easy to revert.

I'd be surprised if this is in fact the case - you're iterating through lines in Python checking them one at a time. The original method did the whole search in a Python stdlib package that I would expect to be essentially compiled C code. Not only am I skeptical of the optimization, but I'm also skeptical this is part of the process that really contributes much to the overhead of running jobs - for these reasons this is seems like a premature optimization and adds complexity I don't think we need - I think it should be reverted if that is okay.

bernt-matthias commented:
> Can we just drop that addition? How about drop that addition and we will backport this

Don't know about backporting. I'm open for this. Could revert to the old matching (non line wise). 

> I'll open a PR against dev that adds a new field on Job like failure_reason so we don't have to modify the input and output detected? 

That's a great idea -- much better than editing output. From my work on job runners I remember that there are already job states. Couldn't those be used/extended (https://github.com/galaxyproject/galaxy/blob/b9d07aed22c59c2aeb92e30d18953bc6fecea0a6/lib/galaxy/model/__init__.py#L543). 

The main part of the PR is the changed logic of how errors are treated: check for OOM even if an exit code is present. I would suggest to adopt the idea .. maybe parts of this PR  (maybe I could rebase this then to your branch) or a reimplementation.

jmchilton commented:
Closing in lieu of #7095 - I hope that it is okay. It contains all these commits with the authorship in tact.
